{
  "metadata": {
    "kernelspec": {
      "name": "python",
      "display_name": "Python (Pyodide)",
      "language": "python"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "python",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.8"
    }
  },
  "nbformat_minor": 4,
  "nbformat": 4,
  "cells": [
    {
      "cell_type": "markdown",
      "source": "# Tuần 2(12/9-18/9)\n\n# Train model theo các video hướng dẫn trên Youtube\n\n## **Thực hiện train model theo video *Real Time Sign Language Detection with Tensorflow Object Detection and Python | Deep Learning SSD* của *Nicholas Renotte***\n\nLink Youtube: [Real Time Sign Language Detection with Tensorflow Object Detection and Python | Deep Learning SSD](https://www.youtube.com/watch?v=pDXdlXlaCco)\n\nLink Github: [RealTimeObjectDetection](https://github.com/nicknochnack/RealTimeObjectDetection)\n\n- Cài đặt Jupyter Notebook theo link hướng dẫn: [Jupyter Notebook là gì](https://200lab.io/blog/jupyter-notebook-la-gi/)\n    \n    Sau khi cài đặt Python version 3.12 và Jupyter Notebook, khởi chạy Jupyter Notebook theo hướng dẫn thì phát hiện máy không thể mở được [Localhost](http://Localhost) của Jupyter Notebook.\n    \n    Khi sao chép link [Localhost](http://Localhost) của Jupyter Notebook từ Command Prompt và dán lên thanh tìm kiếm của chrome thì ra trang Localhost của Jupyter Notebook với nhiều folder và file lạ, không thể thực hiện giống với video, không thể upload folder vào.\n    \n\nSau khi khởi chạy Jupyter Notebook thất bại, chuyển sang video hướng dẫn khác. \n\n## **Thực hiện train model theo danh sách video *Sign Language Detection Using YOLO v5 | Computer Vision* của *DSwithBappy***\n\nLink Youtube: [Sign Language Detection Using YOLO v5 | Computer Vision](https://www.youtube.com/playlist?list=PLkz_y24mlSJYWpwFbU8fyaBSwihoVHiJz)\n\nLink Github: [Sign-Language-Generation-From-Video-using-YOLOV5](https://github.com/entbappy/Sign-Language-Generation-From-Video-using-YOLOV5)\n\nThực hiện mở file *Sign_language_Generation_Using_YOLO_v5.ipynb* trong Github lên Google Colab để thực hiện việc train model nhận diện. Link Google Colab: [Google Collab](https://colab.research.google.com/drive/1O364FjeYc3cX-fnNwXgIPBFuY-gwUmf5?authuser=1)\n\nSau khi thực hiện theo video hướng dẫn thì trong quá trình thực hiện gặp phải lỗi ở phần dưới đây:\n\n**%cp /content/yolov5/runs/train/yolov5s_results/weights/best.pt /content/gdrive/MyDrive/Research/SignLanguageDetection/model**\n\nSau khi khởi chạy thì chương trình thông báo lỗi không tìm thấy *Research/SignLanguageDetection/model* trong Google Drive. Sau khi phát hiện lỗi thì thêm folder Research/SignLanguageDetection/model vào Google Drive và chạy lại đoạn code trên. Sau khi chạy lại thì em thu được model [best.pt](http://best.pt) ở folder model trong folder SignLanguageDetection.\n\nChạy wecam nhận diện theo video hướng dẫn, khi xem video thì phát hiện còn thiếu cài đặt Anaconda. Thực hiện cài đặt Anaconda vào máy. Ngoài ra Github trong video có thêm folder yolov5 mà khi em nhấn vào đường dẫn Github thì không có nó. Em tải thêm folder yolov5 từ Github(link Github: https://github.com/ultralytics/yolov5 )vào folder Sign-Language-Generation-From-Video-using-YOLOV5. Sau khi khởi chạy file [*run.py*](http://run.py) theo hướng dẫn của video thì chương trình thông báo lỗi sau đây: \n\n**Loi@DESKTOP-9V0PM3A MINGW64 ~/Documents/TestCV/Sign-Language-Generation-From-Video-using-YOLOV5/yolov5 (master)\n$ python [run.py](http://run.py/)\ndetect: weights=['[best.pt](http://best.pt/)'], source=0, data=data\\coco128.yaml, imgsz=[416, 416], conf_thres=0.5, iou_thres=0.45, max_det=1000, device=, view_img=False, save_txt=False, save_format=0, save_csv=False, save_conf=False, save_crop=False, nosave=False, classes=None, agnostic_nms=False, augment=False, visualize=False, update=False, project=runs\\detect, name=exp, exist_ok=False, line_thickness=3, hide_labels=False, hide_conf=False, half=False, dnn=False, vid_stride=1\nYOLOv5  v7.0-366-gf7322921 Python-3.9.19 torch-2.4.1+cpu CPU**\n\n**Traceback (most recent call last):\nFile \"C:\\Users\\Loi\\Documents\\TestCV\\Sign-Language-Generation-From-Video-using-YOLOV5\\yolov5\\[detect.py](http://detect.py/)\", line 437, in <module>\nmain(opt)\nFile \"C:\\Users\\Loi\\Documents\\TestCV\\Sign-Language-Generation-From-Video-using-YOLOV5\\yolov5\\[detect.py](http://detect.py/)\", line 432, in main\nrun(**vars(opt))\nFile \"C:\\Users\\Loi\\Documents\\TestCV\\Sign-Language-Generation-From-Video-using-YOLOV5\\env\\lib\\site-packages\\torch\\utils\\_contextlib.py\", line 116, in decorate_context\nreturn func(*args, **kwargs)\nFile \"C:\\Users\\Loi\\Documents\\TestCV\\Sign-Language-Generation-From-Video-using-YOLOV5\\yolov5\\[detect.py](http://detect.py/)\", line 166, in run\nmodel = DetectMultiBackend(weights, device=device, dnn=dnn, data=data, fp16=half)\nFile \"C:\\Users\\Loi\\Documents\\TestCV\\Sign-Language-Generation-From-Video-using-YOLOV5\\yolov5\\models\\[common.py](http://common.py/)\", line 489, in init\nmodel = attempt_load(weights if isinstance(weights, list) else w, device=device, inplace=True, fuse=fuse)**\n\n**File \"C:\\Users\\Loi\\Documents\\TestCV\\Sign-Language-Generation-From-Video-using-YOLOV5\\yolov5\\models\\[experimental.py](http://experimental.py/)\", line 98, in attempt_load\nckpt = torch.load(attempt_download(w), map_location=\"cpu\")  # load\nFile \"C:\\Users\\Loi\\Documents\\TestCV\\Sign-Language-Generation-From-Video-using-YOLOV5\\env\\lib\\site-packages\\ultralytics\\utils\\[patches.py](http://patches.py/)\", line 86, in torch_load\nreturn _torch_load(*args, **kwargs)\nFile \"C:\\Users\\Loi\\Documents\\TestCV\\Sign-Language-Generation-From-Video-using-YOLOV5\\env\\lib\\site-packages\\torch\\[serialization.py](http://serialization.py/)\", line 1097, in load\nreturn _load(\nFile \"C:\\Users\\Loi\\Documents\\TestCV\\Sign-Language-Generation-From-Video-using-YOLOV5\\env\\lib\\site-packages\\torch\\[serialization.py](http://serialization.py/)\", line 1525, in _load\nresult = unpickler.load()\nFile \"C:\\Users\\Loi\\Documents\\TestCV\\Sign-Language-Generation-From-Video-using-YOLOV5\\env\\lib\\[pathlib.py](http://pathlib.py/)\", line 1084, in new\nraise NotImplementedError(\"cannot instantiate %r on your system\"\nNotImplementedError: cannot instantiate 'PosixPath' on your system\n(C:\\Users\\Loi\\Documents11:16:38estCV\\Sign-Language-Generation-From-Video-using-YOLOV5v)\nLoi@DESKTOP-9V0PM3A MINGW64 ~/Documents/TestCV/Sign-Language-Generation-From-Video-using-YOLOV5/yolov5 (master)\n$ python [run.py](http://run.py/)\ndetect: weights=['[best.pt](http://best.pt/)'], source=0, data=data\\coco128.yaml, imgsz=[416, 416], conf_thres=0.5, iou_thres=0.45, max_det=1000, device=, view_img=False, save_txt=False, save_format=0, save_csv=False, save_conf=False, save_crop=False, nosave=False, classes=None, agnostic_nms=False, augment=False, visualize=False, update=False, project=runs\\detect, name=exp, exist_ok=False, line_thickness=3, hide_labels=False, hide_conf=False, half=False, dnn=False, vid_stride=1\nYOLOv5  v7.0-366-gf7322921 Python-3.9.19 torch-2.4.1+cpu CPU**\n\n**Traceback (most recent call last):\nFile \"C:\\Users\\Loi\\Documents\\TestCV\\Sign-Language-Generation-From-Video-using-YOLOV5\\yolov5\\[detect.py](http://detect.py/)\", line 437, in <module>\nmain(opt)\nFile \"C:\\Users\\Loi\\Documents\\TestCV\\Sign-Language-Generation-From-Video-using-YOLOV5\\yolov5\\[detect.py](http://detect.py/)\", line 432, in main\nrun(**vars(opt))\nFile \"C:\\Users\\Loi\\Documents\\TestCV\\Sign-Language-Generation-From-Video-using-YOLOV5\\env\\lib\\site-packages\\torch\\utils\\_contextlib.py\", line 116, in decorate_context\nreturn func(*args, **kwargs)\nFile \"C:\\Users\\Loi\\Documents\\TestCV\\Sign-Language-Generation-From-Video-using-YOLOV5\\yolov5\\[detect.py](http://detect.py/)\", line 166, in run\nmodel = DetectMultiBackend(weights, device=device, dnn=dnn, data=data, fp16=half)\nFile \"C:\\Users\\Loi\\Documents\\TestCV\\Sign-Language-Generation-From-Video-using-YOLOV5\\yolov5\\models\\[common.py](http://common.py/)\", line 489, in init\nmodel = attempt_load(weights if isinstance(weights, list) else w, device=device, inplace=True, fuse=fuse)**\n\n**File \"C:\\Users\\Loi\\Documents\\TestCV\\Sign-Language-Generation-From-Video-using-YOLOV5\\yolov5\\models\\[experimental.py](http://experimental.py/)\", line 98, in attempt_load\nckpt = torch.load(attempt_download(w), map_location=\"cpu\")  # load\nFile \"C:\\Users\\Loi\\Documents\\TestCV\\Sign-Language-Generation-From-Video-using-YOLOV5\\env\\lib\\site-packages\\ultralytics\\utils\\[patches.py](http://patches.py/)\", line 86, in torch_load\nreturn _torch_load(*args, **kwargs)\nFile \"C:\\Users\\Loi\\Documents\\TestCV\\Sign-Language-Generation-From-Video-using-YOLOV5\\env\\lib\\site-packages\\torch\\[serialization.py](http://serialization.py/)\", line 1097, in load\nreturn _load(\nFile \"C:\\Users\\Loi\\Documents\\TestCV\\Sign-Language-Generation-From-Video-using-YOLOV5\\env\\lib\\site-packages\\torch\\[serialization.py](http://serialization.py/)\", line 1525, in _load\nresult = unpickler.load()\nFile \"C:\\Users\\Loi\\Documents\\TestCV\\Sign-Language-Generation-From-Video-using-YOLOV5\\env\\lib\\[pathlib.py](http://pathlib.py/)\", line 1084, in new\nraise NotImplementedError(\"cannot instantiate %r on your system\"\nNotImplementedError: cannot instantiate 'PosixPath' on your system**\n<br>Lỗi gặp phải xuất phát từ việc không thể khởi tạo `PosixPath` trên hệ thống Windows. Điều này xảy ra do code đang cố sử dụng `PosixPath` (dành cho hệ điều hành Unix) trong khi đang sử dụng hệ điều hành Windows, vốn dùng `WindowsPath`.\n\n## **Thực hiện train model theo video *Chapter 50: Yolov8 for custom dataset ASL Sign Language WhatsApp +91-8909603708 for Research Project* của *[Code with Kristi](https://www.youtube.com/@codewithkristi)***\n\n[Link Youtube: https://www.youtube.com/watch?v=FU1nqyhklM8](https://www.notion.so/Link-Youtube-https-www-youtube-com-watch-v-FU1nqyhklM8-10d92668930c80e4ac32f6d233c1dfef?pvs=21)\n\n[Link Github: https://github.com/sachin365123/Yolov8-for-custom-dataset-ASL-Sign-Language](https://www.notion.so/Link-Github-https-github-com-sachin365123-Yolov8-for-custom-dataset-ASL-Sign-Language-10d92668930c8020a974f1c5187ac1e4?pvs=21)\n\nThực hiện mở file *Yolov8_ASL_Sign_Language.ipynb* trong Github lên Google Colab để thực hiện việc train model nhận diện. \n\nSau khi thực hiện theo video hướng dẫn thì trong quá trình thực hiện gặp phải lỗi ở phần dưới đây:\n\n**!mkdir /content/Yolov8SignLanguage-3/datasets\n%cd /content/Yolov8SignLanguage-3/datasets\n\nfrom roboflow import Roboflow\nrf = Roboflow(api_key=\"U9X00mIoCIlSYdyte6NZ\")\nproject = rf.workspace(\"invertis\").project(\"yolov8signlanguage\")\ndataset = project.version(3).download(\"yolov8\")**\n\nEm gặp phải lỗi api_key của dataset Roboflow nên em tìm nguồn dataset khác ở Roboflow để thay thế. Nguồn dataset American Sign Language tìm được là ***American Sign Language Letters Computer Vision Project*** của ***David Lee,*** Link Roboflow: ****https://universe.roboflow.com/david-lee-d0rhs/american-sign-language-letters.\n\nSau khi cài đặt dataset để train model từ nguồn Roboflow thì em lại gặp tiếp lỗi ở phần code dưới đây:\n\n**!yolo task=detect mode=train model=/content/yolov8s.pt data=/content/American-Sign-Language-Letters-6/data.yaml epochs=30 imgsz=640**\n\n***FileNotFoundError:\nDataset '/content/American-Sign-Language-Letters-6/data.yaml' images not found ⚠️, missing path '/content/American-Sign-Language-Letters-6/American-Sign-Language-Letters-6/valid/images'\nNote dataset download directory is '/content/datasets'. You can update this in '/root/.config/Ultralytics/settings.yaml'***\n\nSau khi không tìm được cách xử lý lỗi trên thì em lại tìm video khác.\n\n## **Thực hiện train model theo video *Train YOLOv8 on Custom Dataset | Sign Language Alphabets Detection and Recognition using YOLOv8* của *[Muhammad Moin](https://www.youtube.com/@muhammadmoinfaisal)***\n\n[Link Youtube: https://www.youtube.com/watch?v=-UoSr9Z_Bg0](https://www.notion.so/Link-Youtube-https-www-youtube-com-watch-v-UoSr9Z_Bg0-10d92668930c808a89a3e857ca68d8b0?pvs=21)\n\n[Link Github: https://github.com/MuhammadMoinFaisal/Sign-Language-Alphabets-Detection-and-Recongition-using-YOLOv8](https://www.notion.so/Link-Github-https-github-com-MuhammadMoinFaisal-Sign-Language-Alphabets-Detection-and-Recongition-10d92668930c80879ab0d37e5eb6320d?pvs=21)\n\nThực hiện mở file *Sign_Language_Alphabets_Detection_and_Recognition_using_YOLOv8.ipynb* trong Github lên Google Colab để thực hiện việc train model nhận diện. Link Google Colab: https://colab.research.google.com/drive/1iwUa4TdjQTMgU8gRdNxMtx_qYEXAnfs6?authuser=1#scrollTo=SA3S___qj88u\n\nSau khi thực hiện theo video hướng dẫn thì trong quá trình thực hiện lại gặp phải lỗi ở phần dưới đây:\n\n**%cd {HOME}**\n\n**!yolo task=detect mode=train [model=yolov8l.pt](http://model=yolov8l.pt/) data={dataset.location}/data.yaml epochs=50 imgsz=800**\n\n***FileNotFoundError:\nDataset '/content/American-Sign-Language-Letters-6/data.yaml' images not found ⚠️, missing path '/content/American-Sign-Language-Letters-6/American-Sign-Language-Letters-6/valid/images'\nNote dataset download directory is '/content/datasets'. You can update this in '/root/.config/Ultralytics/settings.yaml'***\n\nSau khi không tìm được cách xử lý lỗi trên thì em lại tìm video khác.\n\n## **Thực hiện train model theo video *Real Time Sign Language Detection with YOLO-NAS and Webcam : Step by Step Tutorial | Custom Dataset* của *[Muhammad Moin](https://www.youtube.com/@muhammadmoinfaisal)***\n\nLink Youtube: [Real Time Sign Language Detection with YOLO-NAS and Webcam : Step by Step Tutorial | Custom Dataset](https://www.youtube.com/watch?v=1FOawpuQflg)\n\nLink Github: [YOLO-NAS Sign Language Detection](https://github.com/MuhammadMoinFaisal/Computervisionprojects/tree/main/YOLO-NAS%20Sign%20Language%20Detection)\n\nThực hiện mở file *Train_YOLONAS_Custom_Dataset_Sign_Language_Complete.ipynb* trong Github lên Google Colab để thực hiện việc train model nhận diện. Link Google Colab: https://colab.research.google.com/drive/1vP6rsiizdUzHYrmFKvwBPdxy35Dyh1NX?authuser=1#scrollTo=63PxbjJi_4jr\n\nSau khi thực hiện theo video hướng dẫn thì trong quá trình thực hiện lại gặp phải lỗi ở phần dưới đây:\n\n**from super_gradients.training import Trainer\nfrom super_gradients.training import dataloaders\nfrom super_gradients.training.dataloaders.dataloaders import coco_detection_yolo_format_train, coco_detection_yolo_format_val**\n\n***--------------------------------------------------------------------------***\n\n***ModuleNotFoundError                       Traceback (most recent call last)***\n\n***<ipython-input-2-0200189a802d> in <cell line: 1>()***\n\n***---> 1 from super_gradients.training import Trainer***\n\n  ***2 from super_gradients.training import dataloaders***\n\n  ***3 from super_gradients.training.dataloaders.dataloaders import coco_detection_yolo_format_train, coco_detection_yolo_format_val***\n\n***ModuleNotFoundError: No module named 'super_gradients'***\n\n***--------------------------------------------------------------------------***\n\n***NOTE: If your import is failing due to a missing package, you can***\n\n***manually install dependencies using either !pip or !apt.***\n\n***To view examples of installing some common dependencies, click the***\n\n***\"Open Examples\" button below.***\n\n***--------------------------------------------------------------------------***\n\nSau khi không tìm được cách xử lý lỗi trên thì em lại tìm video khác.\n\n## **Thực hiện train model theo video *Fire Detection YOLOv8 with Python (real time & custom data)* của *[University of The Future](https://www.youtube.com/@universityofthefuture2077)***\n\nLink Youtube: [Fire Detection YOLOv8 with Python (real time & custom data)](https://www.youtube.com/watch?v=-RDeVPHipZU)\n\nLink Github: [yolov8-on-fire-colab](https://github.com/mcdominik/machine_learning/tree/master/yolov8-on-fire-colab)\n\nThực hiện mở file *yolo8_on_fire.ipynb* trong Github lên Google Colab để thực hiện việc train model nhận diện. Link Google Colab: https://colab.research.google.com/drive/1wW4-CirEYCxmfMrAdO7NHduNbA-56Ru0?authuser=1\n\nSau khi thực hiện theo video hướng dẫn thì trong quá trình thực hiện lại gặp phải lỗi ở phần dưới đây:\n\n**!yolo task=detect mode=train model=yolov8s.pt data=/content/American-Sign-Language-Letters-6/data.yaml epochs=10 imgsz=640**\n\n***FileNotFoundError:\nDataset '/content/American-Sign-Language-Letters-6/data.yaml' images not found ⚠️, missing path '/content/American-Sign-Language-Letters-6/American-Sign-Language-Letters-6/valid/images'\nNote dataset download directory is '/content/datasets'. You can update this in '/root/.config/Ultralytics/settings.yaml'***\n\nSau đó thì em thử thay dataset khác từ Roboflow là ***sign language Computer Vision Project*** của ***Roboflow 100***. Dù vậy sau khi thay vẫn gặp lỗi tương tự là \n\n***FileNotFoundError:***\n\n***Dataset '/content/Sign-Language-2/data.yaml' images not found ⚠️, missing path '/content/Sign-Language-2/Sign-Language-2/valid/images'\nNote dataset download directory is '/content/datasets'. You can update this in '/root/.config/Ultralytics/settings.yaml'***\n\nSau một thời gian thì em đã tìm ra lỗi nằm ở đâu sau khi thêm /content vào hai dòng train và val của data.yaml thì đã chạy train thành công. Sau khi train xong cỡ 10 phút thì thu được model best.pt rồi làm tiếp theo video hướng dẫn. Theo hướng dẫn, em tạo thêm file run.py trên Visual Studio để chạy wecam thì khởi chạy thành công.",
      "metadata": {}
    },
    {
      "cell_type": "code",
      "source": "",
      "metadata": {
        "trusted": true
      },
      "outputs": [],
      "execution_count": null
    }
  ]
}